src/mono/mono/mini/mini-s390x.c

```diff
@@ -2212,6 +2659,227 @@ mono_arch_lowering_pass (MonoCompile *cfg, MonoBasicBlock *bb)
+				break;
+			case CMP_LE_UN:
+			case CMP_LE:
+				temp = ins->sreg1;
+				ins->sreg1 = ins->sreg2;
+				ins->sreg2 = temp;
+			case CMP_GE_UN:
+			case CMP_GE:
+				ins->opcode = GINT_TO_OPCODE (simd_type_to_ge_fp_op (GTMREG_TO_INT (ins->inst_c1)));
+				break;
+			default:
+				g_assert_not_reached ();
+				break;
+			}
+			break;
+		}
+		case OP_XCOMPARE:{
+			switch (ins->inst_c0){
+			case CMP_EQ:
+				ins->opcode = GINT_TO_OPCODE (simd_type_to_comp_op (GTMREG_TO_INT (ins->inst_c1)));
+				break;
+			case CMP_LT:
+				temp = ins->sreg1;
+				ins->sreg1 = ins->sreg2;
+				ins->sreg2 = temp;
+			case CMP_GT:
+				ins->opcode = GINT_TO_OPCODE (simd_type_to_gt_op (GTMREG_TO_INT (ins->inst_c1)));
+				break;
+			case CMP_LT_UN:
+				temp = ins->sreg1;
+				ins->sreg1 = ins->sreg2;
+				ins->sreg2 = temp;
+			case CMP_GT_UN:
+				ins->opcode = GINT_TO_OPCODE (simd_type_to_gt_op (GTMREG_TO_INT (ins->inst_c1)));
+				break;
+			case CMP_GE:
+			case CMP_GE_UN:
+				temp = ins->sreg1;
+				ins->sreg1 = ins->sreg2;
+				ins->sreg2 = temp;
+			case CMP_LE:
+			case CMP_LE_UN:{
+				NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_gt_op (GTMREG_TO_INT (ins->inst_c1))), ins->dreg, ins->sreg1, ins->sreg2);
+				NEW_SIMD_INS (cfg, ins, temp_ins, OP_S390_VNO, ins->dreg, ins->dreg, ins->dreg);
+				NULLIFY_INS (ins);
+				break;
+			}
+			default:
+				g_assert_not_reached ();
+				break;
+			}
+			break;
+		}
+		case OP_S390_XCOMPARE_XEXTRACT:{
+			switch (type_is_float(GTMREG_TO_INT(ins->inst_c1))){
+			case OP_XCOMPARE:{
+				switch (ins->inst_c0 >> 4){
+                       		case CMP_EQ:
+					NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_comp_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                       		        break;
+                       		case CMP_LT:
+                       		case CMP_LT_UN:
+                       		        temp = ins->sreg1;
+                       		        ins->sreg1 = ins->sreg2;
+                       		        ins->sreg2 = temp;
+                       		case CMP_GT:
+				case CMP_GT_UN:
+					NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_gt_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                       		        break;
+                       		case CMP_GE:
+                       		case CMP_GE_UN:
+                       		        temp = ins->sreg1;
+                       		        ins->sreg1 = ins->sreg2;
+                       		        ins->sreg2 = temp;
+                       		case CMP_LE:
+                       		case CMP_LE_UN:{
+                       		        NEW_SIMD_INS (cfg, ins, temp_ins, GINT_TO_OPCODE (simd_type_to_gt_any_all_op (GTMREG_TO_INT (ins->inst_c1))), s390_vr16, ins->sreg1, ins->sreg2);
+                       		        NEW_SIMD_INS (cfg, ins, temp_ins, OP_S390_VNO, s390_vr16, s390_vr16, s390_vr16);
```

Comment:
This looks wrong.   Nobody will ever look at the vr16 contents here, so negating them doesn't help.  Instead, the user of the *condition code* will have to use the inverted CC mask, right?